--- a/src/nemo-main-application.c	2026-08-15 20:21:54.018404481 +0000
+++ b/src/nemo-main-application.c	2026-08-15 20:21:54.020445743 +0000
@@ -463,13 +463,55 @@
 }
 
 static void
+navigate_existing_window (NemoMainApplication *application,
+                           GFile *location,
+                           GdkScreen *screen,
+                           const char *geometry)
+{
+    GList *list_copy;
+    GList *l;
+    gchar *uri;
+
+    list_copy = g_list_copy (gtk_application_get_windows (GTK_APPLICATION (&application->parent)));
+    for (l = list_copy; l != NULL; l = l->next) {
+        if (NEMO_IS_WINDOW (l->data)) {
+            NemoWindow *window;
+
+            window = NEMO_WINDOW (l->data);
+
+            uri = g_file_get_uri (location);
+            g_debug ("Navigating existing window to uri %s\n", uri);
+            nemo_window_go_to (window, location);
+            g_free (uri);
+
+            /* Don't use `gtk_window_present()`, as the window manager will ignore this window's focus request and try
+             * to just mark it urgent instead (flashing in the window list for example). */
+            if (eel_check_is_wayland ()) {
+                gtk_window_present (GTK_WINDOW (window));
+            } else {
+                gtk_window_present_with_time (GTK_WINDOW (window),
+                                              gdk_x11_get_server_time (gtk_widget_get_window (GTK_WIDGET (window))));
+            }
+
+          break;
+        }
+    }
+    if (l == NULL) {
+        /* no existing window was found, so open a new window */
+        open_window (application, location, screen, geometry);
+    }
+    g_list_free (list_copy);
+}
+
+static void
 open_windows (NemoMainApplication *application,
 	      GFile **files,
 	      gint n_files,
 	      GdkScreen *screen,
 	      const char *geometry,
 	      gboolean open_in_tabs,
-	      gboolean open_in_existing_window)
+	      gboolean open_in_existing_window,
+	      gboolean open_in_self_window)
 {
 	gint i;
 
@@ -477,7 +519,10 @@
 		/* Open a window pointing at the default location. */
 		open_window (application, NULL, screen, geometry);
 	} else {
-		if (open_in_existing_window) {
+		if (open_in_self_window) {
+			/* Navigate an existing window to the (single) requested location */
+			navigate_existing_window (application, files[0], screen, geometry);
+		} else if (open_in_existing_window) {
 			/* Open one tab at each requested location in an existing window */
 			open_tabs_in_existing_window (application, files, n_files, screen, geometry);
 		} else if (open_in_tabs) {
@@ -533,11 +578,12 @@
 	gboolean open_in_tabs = FALSE;
 	gchar *geometry = NULL;
 	gboolean open_in_existing_window = strcmp (options, "EXISTING_WINDOW") == 0;
+	gboolean open_in_self_window = strcmp (options, "SELF_WINDOW") == 0;
 	const char splitter = '=';
 
 	g_debug ("Open called on the GApplication instance; %d files", n_files);
 
-	if (!open_in_existing_window) {
+	if (!open_in_existing_window && !open_in_self_window) {
 		/* Check if local command line passed --geometry or --tabs */
 		if (strlen (options) > 0) {
 			gchar** split_options = g_strsplit (options, &splitter, 2);
@@ -550,13 +596,14 @@
 	}
 
 	DEBUG ("Open called on the GApplication instance; %d files, open in tabs: %s, geometry: '%s',"
-           "open in existing window: %s",
+           "open in existing window: %s, navigate self: %s",
            n_files,
            open_in_tabs ? "yes" : "no",
            geometry ? geometry : "none",
-           open_in_existing_window ? "yes" : "no");
+           open_in_existing_window ? "yes" : "no",
+           open_in_self_window ? "yes" : "no");
 
-	open_windows (self, files, n_files, gdk_screen_get_default (), geometry, open_in_tabs, open_in_existing_window);
+	open_windows (self, files, n_files, gdk_screen_get_default (), geometry, open_in_tabs, open_in_existing_window, open_in_self_window);
 
     g_clear_pointer (&geometry, g_free);
 }
@@ -653,6 +700,7 @@
 	gboolean browser = FALSE;
 	gboolean open_in_tabs = FALSE;
 	gboolean open_in_existing_window = FALSE;
+	gboolean open_in_self_window = FALSE;
 	gboolean kill_shell = FALSE;
 	gboolean no_default_window = FALSE;
     gboolean no_desktop_ignored = FALSE;
@@ -683,6 +731,8 @@
 		  N_("Open URIs in tabs."), NULL },
 		{ "existing-window", 0, 0, G_OPTION_ARG_NONE, &open_in_existing_window,
 		  N_("Open URIs in an existing window."), NULL },
+		{ "self", 0, 0, G_OPTION_ARG_NONE, &open_in_self_window,
+		  N_("Navigate an existing window to URI, instead of opening a new tab or window."), NULL },
 		{ "fix-cache", '\0', 0, G_OPTION_ARG_NONE, &fix_cache,
 		  N_("Repair the user thumbnail cache - this can be useful if you're having trouble with file thumbnails.  Must be run as root"), NULL },
         { "debug", 0, 0, G_OPTION_ARG_NONE, &debug,
@@ -831,7 +881,9 @@
 	/* Invoke "Open" to open in existing window or create new windows */
 	if (len > 0) {
 		gchar* concatOptions = g_malloc0(64);
-		if (open_in_existing_window) {
+		if (open_in_self_window) {
+			g_stpcpy (concatOptions, "SELF_WINDOW");
+		} else if (open_in_existing_window) {
 			g_stpcpy (concatOptions, "EXISTING_WINDOW");
 		} else {
 			if (self->priv->geometry == NULL) {
